2019-08-03-q-and-as-about-nix-for-haskellers.md

```diff
@@ -0,0 +1,692 @@
+Reddit, YouTube, etc), know about its advantages, but haven't actually played
+around with it.  It is mainly a **high-level overview** of how Nix works with
+Haskell.  It is not a hands-on tutorial or walk-through.
+
+The first section of this article is an intro to Nix, followed by answers
+to each of Dmitrii's questions.[^7]
+
+## Intro
+
+You may have heard that Nix is complicated and hard to learn.  In my
+experience, this is true.
+
+It is similar to Haskell.  The learning curve is steep, but once you get the
+hang of it, it is very powerful and flexible.
+
+*What makes it so hard?*
+
+To start off with, Nix refers to two things:
+
+1. A programming language for defining _derivations_
+
+2. A set of command line tools for actually building and installing _derivations_
+
+What is a _derivation_?  You can think of it as something similar
+to a "package" in a distribution like Debian or Fedora.
+
+The Nix language provides a way of defining derivations.  The following code
+shows an example of defining a derivation for the GNU
+[hello](https://www.gnu.org/software/hello/hello.html) package[^1]:
+
+```nix
+stdenv.mkDerivation {
+  name = "hello-2.1.1";
+  builder = ./builder.sh;
+  src = fetchurl {
+    url = ftp://ftp.nluug.nl/pub/gnu/hello/hello-2.1.1.tar.gz;
+    sha256 = "1md7jsfd8pa45z73bz1kszpp01yw6x5ljkjk2hx7wl800any6465";
+  };
+```
+
+As far as the Nix language is concerned, you can think of a derivation as a
+dictionary of key-value pairs.  Some of the keys (and corresponding values) are
+considered "special".  For instance, in the above definition, `builder` is a
+special key that corresponds to a shell script.  The shell script contains
+commands like `./configure && make && make install` that will be used to
+build and install the package.
+
+Nix has a few well-known CLI tools.  In this article I will mostly talk about
+`nix-build` and `nix-shell`.  `nix-build` would take the above derivation and
+actually build it.  `nix-build` uses the "special" keys in the
+above derivation to run shell scripts to build and install the binaries, shared
+objects, header files, and other files associated with the package.[^2]
+
+For instance, if the above derivation is saved in a file called
+`my-first-derivation.nix`, you can build it with the following command:[^3]
+
+```
+$ nix-build ./my-first-derivation.nix
+```
+
+This produces a directory in the Nix store called something like
+`/nix/store/06vykrz1hmxgxir8i74fwjl6r9bb2gpg-hello-2.10`.  This contains, among
+other things, the `hello` binary:
+
+```
+$ /nix/store/06vykrz1hmxgxir8i74fwjl6r9bb2gpg-hello-2.10/bin/hello
+Hello, world!
+```
+
+*That doesn't sound too bad.  Where is the difficulty?*
+
+If this was all there is to Nix, it wouldn't be too difficult.  However, it
+also wouldn't be that helpful.
+
+The other big part of Nix is a huge set of pre-defined derivations and helper
+functions in a repo called [nixpkgs](https://github.com/NixOS/nixpkgs).
```

Comment:
Probably some analogy with Haskell will help here. Is `nixpkgs` is like Hackage to Haskell? Or like `base` standard library?